In [109]:
import pandas as pd

In [111]:
df = pd.read_csv("./titanic.csv")

In [113]:
df

Unnamed: 0  pclass                                             name  \
0              0       1                    Allen, Miss. Elisabeth Walton   
1              1       1                   Allison, Master. Hudson Trevor   
2              2       1                     Allison, Miss. Helen Loraine   
3              3       1             Allison, Mr. Hudson Joshua Creighton   
4              4       1  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...          ...     ...                                              ...   
1304        1304       3                             Zabour, Miss. Hileni   
1305        1305       3                            Zabour, Miss. Thamine   
1306        1306       3                        Zakarian, Mr. Mapriededer   
1307        1307       3                              Zakarian, Mr. Ortin   
1308        1308       3                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest  survived  
0       NaN                     St Louis, MO         1  
1       NaN  Montreal, PQ / Chesterville, ON         1  
2       NaN  Montreal, PQ / Chesterville, ON         0  
3     135.0  Montreal, PQ / Chesterville, ON         0  
4       NaN  Montreal, PQ / Chesterville, ON         0  
...     ...                              ...       ...  
1304  328.0                              NaN         0  
1305    NaN                              NaN         0  
1306  304.0                              NaN         0  
1307    NaN                              NaN         0  
1308    NaN                              NaN         0  

[1309 rows x 15 columns]

In [115]:
from sklearn.datasets import fetch_openml

mnist = fetch_openml('mnist_784', as_frame=False)

In [116]:
X, y = mnist.data, mnist.target

In [117]:
df.describe()

Unnamed: 0       pclass          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean    654.000000     2.294882    29.881135     0.498854     0.385027   
std     378.020061     0.837836    14.413500     1.041658     0.865560   
min       0.000000     1.000000     0.166700     0.000000     0.000000   
25%     327.000000     2.000000    21.000000     0.000000     0.000000   
50%     654.000000     3.000000    28.000000     0.000000     0.000000   
75%     981.000000     3.000000    39.000000     1.000000     0.000000   
max    1308.000000     3.000000    80.000000     8.000000     9.000000   

              fare        body     survived  
count  1308.000000  121.000000  1309.000000  
mean     33.295479  160.809917     0.381971  
std      51.758668   97.696922     0.486055  
min       0.000000    1.000000     0.000000  
25%       7.895800   72.000000     0.000000  
50%      14.454200  155.000000     0.000000  
75%      31.275000  256.000000     1.000000  
max     512.329200  328.000000     1.000000

##### count가 데이터셋의 길이를 보여줌. 총 1309명의 데이터이기 때문에 이를 6:1 비율로 나누면 훈련 세트랑 테스트 세트는 각각 1122개랑 187개가 됨

In [122]:
X_train, X_test, y_train, y_test = X[:1122], X[1122:], y[:1122], y[1122:]

In [124]:
y_train_1 = (y_train == '1')
y_test_1 = (y_test == '1')

In [126]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(random_state=42)
sgd_clf.fit(X_train, y_train_1)

SGDClassifier(random_state=42)

In [128]:
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix

y_train_pred = cross_val_predict(sgd_clf, X_train, y_train_5, cv=3)
confusion_matrix(y_train_5, y_train_pred)

array([[976,  12],
       [  9, 125]])

### 정확도
##### 정확도: 전체 중 올바르게 맞춘 비율
$976\over 976+12+9+125$=$1101\over 1122$=$0.98$ → 98% 정확도

### 정밀도 vs 재현율
##### 정밀도:생존자라고 예측한 사람 중 실제로 생존한 사람의 비율
$TP\over FP+TP$
>$125\over 12+125 \\ =0.912$ → 91.2%
<br>

##### 재현율:실제 생존자 중에서 생존자라고 올바르게 예측한 비율
$TP\over FN+TP$
> $125\over 9+125 \\ =0.925$ → 92.5%

##### 재현율이 더 중요한 이유: 
##### 생존자가 아닌 사람, 즉 사망자를 생존자라 판별하는 상황이 생기더라도 생존자를 놓치지 않는게 가장 결정적이기 때문!